File: app/src/main/java/com/example/autoflow/data/WorkflowDao.java
package com.example.autoflow.data;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List; // Import List

@Dao
public interface WorkflowDao {

    @Insert
    long insert(WorkflowEntity workflowEntity);

    @Update
    int update(WorkflowEntity workflowEntity);

    @Delete
    void delete(WorkflowEntity workflowEntity);

    @Query("DELETE FROM workflows WHERE id = :workflowId")
    int deleteById(long workflowId);

    @Query("DELETE FROM workflows")
    void deleteAll();

    @Query("SELECT * FROM workflows")
    List<WorkflowEntity> getAllWorkflowsSync();

    @Query("SELECT * FROM workflows WHERE is_enabled = 1")
    List<WorkflowEntity> getEnabledWorkflowsSync();

    @Query("SELECT * FROM workflows WHERE id = :id")
    WorkflowEntity getByIdSync(long id);

    @Query("UPDATE workflows SET is_enabled = :enabled WHERE id = :workflowId")
    int updateEnabled(long workflowId, boolean enabled);

    @Query("SELECT COUNT(*) FROM workflows")
    int getCount();

    // LiveData versions for reactive programming (optional)
    @Query("SELECT * FROM workflows")
    LiveData<List<WorkflowEntity>> getAllWorkflows();

    @Query("SELECT * FROM workflows WHERE is_enabled = 1")
    LiveData<List<WorkflowEntity>> getEnabledWorkflows();

    @Query("SELECT * FROM workflows WHERE id = :id")
    LiveData<WorkflowEntity> getById(long id);

}

File: app/src/main/java/com/example/autoflow/data/WorkflowDao.kt
package com.example.autoflow.data

import androidx.lifecycle.LiveData
import androidx.room.*

@Dao
interface WorkflowDao {
    // ✅ INSERT METHODS
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(workflowEntity: WorkflowEntity): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertSync(workflow: WorkflowEntity): Long

    // ✅ UPDATE METHODS
    @Update
    fun update(workflowEntity: WorkflowEntity): Int

    @Update
    fun updateSync(workflow: WorkflowEntity)

    @Query("UPDATE workflows SET is_enabled = :enabled WHERE id = :workflowId")
    fun updateEnabled(workflowId: Long, enabled: Boolean): Int

    @Query("UPDATE workflows SET is_enabled = :enabled WHERE id = :workflowId")
    fun updateEnabledSync(workflowId: Long, enabled: Boolean)

    // ✅ DELETE METHODS
    @Delete
    fun delete(workflowEntity: WorkflowEntity)

    @Query("DELETE FROM workflows WHERE id = :workflowId")
    fun deleteById(workflowId: Long): Int

    @Query("DELETE FROM workflows WHERE id = :workflowId")
    fun deleteByIdSync(workflowId: Long)

    @Query("DELETE FROM workflows")
    fun deleteAll()

    // ✅ SYNCHRONOUS QUERIES
    @Query("SELECT * FROM workflows ORDER BY created_at DESC")
    fun getAllWorkflowsSync(): List<WorkflowEntity>

    @Query("SELECT * FROM workflows WHERE is_enabled = 1")
    fun getEnabledWorkflowsSync(): List<WorkflowEntity>

    @Query("SELECT * FROM workflows WHERE id = :id")
    fun getByIdSync(id: Long): WorkflowEntity?

    @Query("SELECT COUNT(*) FROM workflows")
    fun getCount(): Int

    @Query("SELECT * FROM workflows")
    fun getAllSyncBlocking(): List<WorkflowEntity>

    // ✅ SUSPEND QUERIES (for coroutines)
    @Query("SELECT * FROM workflows")
    suspend fun getAllSync(): List<WorkflowEntity>

    @Query("SELECT * FROM workflows WHERE is_enabled = 1")
    suspend fun getAllEnabledSync(): List<WorkflowEntity>

    @Query("SELECT * FROM workflows WHERE is_enabled = 1")
    suspend fun getActiveWorkflows(): List<WorkflowEntity>

    // ✅ LIVEDATA QUERIES (for UI observations)
    @Query("SELECT * FROM workflows ORDER BY created_at DESC")
    fun getAll(): LiveData<List<WorkflowEntity>>

    @Query("SELECT * FROM workflows ORDER BY created_at DESC")
    fun getAllWorkflows(): LiveData<List<WorkflowEntity>>

    @Query("SELECT * FROM workflows WHERE is_enabled = 1")
    fun getEnabledWorkflows(): LiveData<List<WorkflowEntity>>

    @Query("SELECT * FROM workflows WHERE id = :id")
    fun getById(id: Long): LiveData<WorkflowEntity?>

    // ✅ SEARCH QUERIES
    @Query("SELECT * FROM workflows WHERE workflow_name LIKE '%' || :query || '%'")
    fun search(query: String): List<WorkflowEntity>
}
